src/modules/storage/LedgerStorage.ts

```diff
@@ -2275,23 +2377,58 @@ export class LedgerStorage extends Storages {
                     .catch((err) => {
                         reject(err);
                     });
+
             });
         }
-        function putPayload(storage: LedgerStorage, block_height: Height, tx_hash: Hash, tx: Transaction) {
+
+        function getProposalVotePeriod(storage: LedgerStorage, proposal_id: string, app_name: string): Promise<any[]> {
+            let sql = `SELECT 
+                    vote_start_height, vote_end_height 
+                    FROM 
+                        proposal
+                    WHERE proposal_id = ? and app_name = ?;`;
+
+            return storage.query(sql, [proposal_id, app_name]);
+        }
+
+        function putPayload(storage: LedgerStorage, block_header: BlockHeader, tx_hash: Hash, tx: Transaction) {
             return new Promise<void>(async (resolve, reject) => {
                 try {
-                    if (tx.payload.length >= 9) {
-                        const header = tx.payload.slice(1, 9);
-                        if (Buffer.compare(Buffer.from(ProposalFeeData.HEADER), header) === 0) {
-                            const buffer = SmartBuffer.fromBuffer(tx.payload);
-                            const proposal_fee = ProposalFeeData.deserialize(buffer);
-                            await save_proposal_fee(storage, block_height, tx_hash, proposal_fee.proposal_id);
-                        } else if (Buffer.compare(Buffer.from(ProposalData.HEADER), header) === 0) {
-                            const buffer = SmartBuffer.fromBuffer(tx.payload);
-                            const proposalData: ProposalData = ProposalData.deserialize(buffer);
-                            await save_proposal_data(storage, block_height, tx_hash, proposalData);
-                        } else if (Buffer.compare(Buffer.from(BallotData.HEADER), header) === 0) {
-                            // TODO
+                    let buffer = SmartBuffer.fromBuffer(tx.payload);
+                    let length = VarInt.toNumber(buffer);
```

Comment:
There was an error in this code, so it was merged by uploading the modified PR before.
https://github.com/bosagora/stoa/pull/462
``` TypeScript

        function putPayload(storage: LedgerStorage, block_header: BlockHeader, tx_hash: Hash, tx: Transaction) {
            return new Promise<void>(async (resolve, reject) => {
                try {
                    if (tx.payload.length >= 9) {
                        const header = tx.payload.slice(1, 9);
                        if (Buffer.compare(Buffer.from(ProposalFeeData.HEADER), header) === 0) {
                            const buffer = SmartBuffer.fromBuffer(tx.payload);
                            const proposal_fee = ProposalFeeData.deserialize(buffer);
                            await save_proposal_fee(storage, block_header.height, tx_hash, proposal_fee.proposal_id);
                        } else if (Buffer.compare(Buffer.from(ProposalData.HEADER), header) === 0) {
                            const buffer = SmartBuffer.fromBuffer(tx.payload);
                            const proposalData: ProposalData = ProposalData.deserialize(buffer);
                            await save_proposal_data(storage, block_header.height, tx_hash, proposalData);
                        } else if (Buffer.compare(Buffer.from(BallotData.HEADER), header) === 0) {
                            const buffer = SmartBuffer.fromBuffer(tx.payload);
                            const ballotData: BallotData = BallotData.deserialize(buffer);
                            const info = await getProposalVotePeriod(
                                storage,
                                ballotData.proposal_id,
                                ballotData.app_name
                            );
                            if (info.length)
                                block_header.height.value >= info[0].vote_start_height &&
                                block_header.height.value <= info[0].vote_end_height
                                    ? await save_ballot_data(storage, block_header, tx_hash, ballotData)
                                    : await save_ballot_data(
                                          storage,
                                          block_header,
                                          tx_hash,
                                          ballotData,
                                          BallotData.REJECT
                                      );
                        }
                    }
                    resolve();
                } catch (err) {
                    logger.error("Failed to put proposal's information:" + err, {
                        operation: Operation.block_sync,
                        height: HeightManager.height.toString(),
                        success: false,
                    });
                    reject(err);
                }
            });
        }
```

Reply:
applied